src/Data/Text/Encoding.hs

```diff
@@ -187,31 +235,311 @@ decodeLatin1 bs = withBS bs $ \fp len -> runST $ do
+        | otherwise = (ndx0, Right $ Utf8ValidState (appendPartialCodePoint 0 $ B.drop ndx0 bs) s)
+
+-- | Validate a 'ByteString' as a contiuation of UTF-8-encoded text.
+--
+-- @validateNextUtf8Chunk chunk s = (n, es)@
+--
+-- This function returns two values:
+--
+-- * The value 'n' indicates the end position of longest prefix of the
+--   'ByteString' that is valid UTF-8-encoded data from the starting
+--   state 's'. If 's' contains an incomplete code point, the input
+--   'ByteString' is considered a continuation. As a result 'n' will be
+--   negative if the code point is still incomplete or is proven to be
+--   invalid.
+--   
+-- * The value 'es' indicates whether the 'ByteString'
+--
+--     * (@Left p@) contains an invalid code point and where the next
+--       (potentially valid) code point begins, so that @p - n@ is the
+--       number of invalid bytes, or
+--     * (@Right s'@) is valid, and all of the remaining bytes starting
+--       at index 'n' are the beginning of an incomplete UTF-8 code
+--       point, and `s'` is the resulting 'Utf8ValidState' value, which
+--       can be used to validate against a following 'ByteString'.
+validateNextUtf8Chunk ::
+#if defined(ASSERTS)
+  HasCallStack =>
 #endif
+  ByteString -> Utf8ValidState -> (Int, Either Int Utf8ValidState)
+validateNextUtf8Chunk bs@(B.length -> len) st@(Utf8ValidState lead s)
+  | len > 0 =
+    let g pos s'
+          -- first things first. let's try to get to the start of the next code point
+          | isUtf8StateIsComplete s' =
+            -- found the beginning of the next code point, hand this off to someone else
+            case validateUtf8Chunk $ B.drop pos bs of
+              (len', mS) -> (pos + len', first (+ pos) mS)
+          -- code point is not complete yet
+          -- walk the rest of the code point until error, complete, or no more data
+          | pos < len =
+            case updateUtf8State (B.index bs pos) s' of
+              -- error
+              Nothing -> (leadPos, Left pos)
+              -- keep going
+              Just s'' -> g (pos + 1) s''
+          -- no more data
+          | otherwise = (leadPos, Right $ Utf8ValidState (appendPartialCodePoint lead bs) s')
+    in g 0 s
+  | otherwise = (leadPos, Right st)
+    where leadPos = -(partUtf8CPLen lead)
+
+-- | Validated UTF-8 data to be converted into a 'Text' value.
+data TextDataStack = TextDataStack
+  { -- | Returns a list of 'Text' and UTF-8-valid 'ByteString' values.
+    dataStack :: [Either Text (Either Word32 ByteString)]
+    -- | Returns total number of UTF-8 valid bytes in the stack.
+  , stackLen :: Int
+  }
+  deriving Show
+
+-- | Empty stack
+emptyStack :: TextDataStack
+emptyStack = TextDataStack [] 0
+
+-- | Push a text value onto the stack
+pushText :: Text -> TextDataStack -> TextDataStack
+pushText t@(Text _ _ tLen) tds@(TextDataStack stack sLen) =
+  if tLen > 0
+  then TextDataStack (Left t : stack) $ sLen + tLen
+  else tds
```

Comment:
This seems to only ever be used to push singletons. Remove it and the `Either Text` from `dataStack`?

Reply:
This module only uses it to push singletons, but the idea goes beyond how it's implemented here. `pushText` is intended to allow the caller to inject any `Text` onto the stack. Instead of having a single replacement character for invalid code point, it could be a series of characters. Of course it doesn't even have to be at an invalid code point. It can be at any point in between chunk decodings.

Reply:
I see. I'm not a fan of this data type  It's going to take more iterations to refine the API. Hence I think it's better to reduce the scope of this PR to just the validation functions, which handles the UTF-8-specific logic. This already makes decoding just a matter of slicing the input strings using the indices provided by the validation functions. We can commit to a higher-level interface in a different proposal.